{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Environment:\n",
    "\n",
    "Login into the GDAV server\n",
    "\n",
    "`$ ssh youruser@IP`\n",
    "\n",
    "Create a directory in your home folder called `compgenomics_ex1`\n",
    "\n",
    "`$ cd javier.fgonzalez`\n",
    "\n",
    "`$ mkdir compgenomics_ex1`\n",
    "\n",
    "and Enter the directory \n",
    "\n",
    "`$ cd compgenomics_ex1`\n",
    "\n",
    "All the files needed for this exercise are copied in the GDAV server at `/home/compgenomics/4proteomes/`. Make sure you can see them and take a few seconds to understand what they contain: \n",
    "\n",
    "```\n",
    "$ ls /home/compgenomics/4proteomes/\n",
    "4proteomes.faa    -> all protein sequences from 4 species: human, elefant, Zebrafish and Ciona intestinalis. \n",
    "G3T0S8_LOXAF.faa  -> protein sequence of the elefant gene called G3T0G8\n",
    "TPH1A_rerio.faa   ->  protein sequence of the Zebrafish gene called TPH1A\n",
    "TPH2_human.faa    -> protein sequence of Human TPH2 \n",
    "scripts/          ->  a directory with ad hoc programs and scripts`\n",
    "```\n",
    "\n",
    "Puedes hacer esto para ver cómo están hechos los fasta: \n",
    "\n",
    "`$ cd ~`\n",
    "\n",
    "`$ head /home/compgenomics/4proteomes/G3T0S8_LOXAF.faa` Con esto puedes ver que las entradas para elefante contienen _LOXAF en el encabezado\n",
    "\n",
    "Las de pez cebra tienen _DANRE\n",
    "\n",
    "Las de humano contienen _HUMAN\n",
    "\n",
    "Probando con greps encontré las de Ciona intestinalis\n",
    "\n",
    "`$ grep 'CIO' /home/compgenomics/4proteomes/4proteomes.faa`\n",
    "\n",
    "Con esto encontré que las de Ciona contienen _CIOIN\n",
    "\n",
    "\n",
    "## Tools needed\n",
    "(already installed in the GDAV server)\n",
    "\n",
    "- BLAST+ \n",
    "\n",
    "# Exercise\n",
    "\n",
    "## Goal 1 (Loxodonta orthologs)\n",
    "\n",
    "Antes de nada vuelve a hacer:\n",
    "\n",
    "` $ cd compgenomics_ex1`\n",
    "\n",
    "Using just BLAST reciprocal searches, could you identify which is the human ortholog of the Loxodonta protein G3T0S8? (remember that the protein sequence of G3T0S8 is available in `/home/compgenomics/4proteomes/G3T0S8_LOXAF.faa`)\n",
    "\n",
    "### Protocol\n",
    "\n",
    "#### 1. Create a BLAST database\n",
    "Make a blast database containing the 4 input proteomes (Human, Danio rerio, Ciona intestinalis, and Loxodonta africana) and name the database as “4proteomes.blastdb’. All proteomes are already merged into a single FASTA file `/home/compgenomics/4proteomes/4proteomes.faa`\n",
    "\n",
    "```\n",
    "$ makeblastdb -dbtype prot -in /home/compgenomics/4proteomes/4proteomes.faa -out 4proteomes.blastdb\n",
    "```\n",
    "\n",
    "You should now see something like this in your exercise home folder: \n",
    "```\n",
    "(base) [test@gdav compgenomics_ex1]$ ls -l\n",
    "total 48604\n",
    "-rw-rw-r--. 1 test test  7434160 oct 15 15:58 4proteomes.blastdb.phr\n",
    "-rw-rw-r--. 1 test test   683920 oct 15 15:58 4proteomes.blastdb.pin\n",
    "-rw-rw-r--. 1 test test 41649539 oct 15 15:58 4proteomes.blastdb.psq\n",
    "```\n",
    "#### 2. Find G3T0S8_LOXAF.faa homologs. \n",
    "\n",
    "Use the `blastp` command to search for all homologs of the G3T0S8 sequence. Use an evalue threshold of 0.001\n",
    "\n",
    "```\n",
    "$ blastp -task blastp -query /home/compgenomics/4proteomes/G3T0S8_LOXAF.faa -db 4proteomes.blastdb -outfmt 6 -evalue 0.001\n",
    "\n",
    "```\n",
    "Es más cómodo guardarlo en un archivo \n",
    "\n",
    "```\n",
    "$ blastp -task blastp -query /home/compgenomics/4proteomes/G3T0S8_LOXAF.faa -db 4proteomes.blastdb -outfmt 6 -evalue 0.001 >> G3T0S8_homologs\n",
    "```\n",
    "\n",
    "\n",
    "#### 3. Answer these questions\n",
    "1. How many homologs of G3T0S8 are in HUMAN?\n",
    "\n",
    "\n",
    "```\n",
    "\n",
    "$ grep '_HUMAN' G3T0S8_homologs | wc -l\n",
    "\n",
    "```\n",
    "\n",
    "Te da 4.\n",
    "\n",
    "2. Which is the closest one?\n",
    "\n",
    "He hecho:\n",
    "\n",
    "```\n",
    "$ grep '_HUMAN' G3T0S8_homologs |sort -g -k 11\n",
    "\n",
    "```\n",
    "-k 11 significa que ordenes según la undécima columna (se empieza a contar por 1)\n",
    "\n",
    "-g significa que hay números exponenciales y que tiene que leerlos bien.\n",
    "\n",
    "\n",
    "El menor e-value es TPH1_HUMAN_TPH1. Ese es el resultado\n",
    "\n",
    "\n",
    "3. Are they orthologs?\n",
    "\n",
    "Hay que hacer blast usando TPH1_HUMAN_TPH1 como query y ver si el mejor match es G3T0S8 (método Best Reciprocal Hits, BRH). En caso afirmativo son ortólogos. En caso negativo puede que sean ortólogos o puede que no (ver apuntes).\n",
    "\n",
    "Lo primero es guardar el gen TPH1 el solo en un fasta:\n",
    "\n",
    "```\n",
    "$ grep -A 1 'TPH1_HUMAN_TPH1' /home/compgenomics/4proteomes/4proteomes.faa >> TPH1_HUMAN.faa\n",
    "```\n",
    "\n",
    "-A 1 lo que hace es meter en el output la línea que va después de la línea en la que encontraste el match (ej si quisieras 3 líneas pones -A 3). Solo queremos 1 porque el match está en el encabezado y toda la secuencia cuenta como una única línea.\n",
    "\n",
    "\n",
    "\n",
    "```\n",
    "$ blastp -task blastp -query TPH1_HUMAN.faa -db 4proteomes.blastdb -outfmt 6 -evalue 0.001 >> TPH1_homologs\n",
    "```\n",
    "\n",
    "```\n",
    "$ grep '_LOXAF' TPH1_homologs |sort -g -k 11\n",
    "\n",
    "```\n",
    "\n",
    "el menor e-value (0) es para el gen G3SPB9_LOXAF_TPH2. Como no ha sido G3T0S8 eso quiere decir que o bien TPH1_HUMAN y G3T0S8 no son ortólogos o bien hay una relación de ortología complicada (ej one to many; many to many, etc.).\n",
    "\n",
    "#### 4. Save homologous sequences\n",
    "\n",
    "Vaya esto ya lo hice antes. Estos son otros comandos para hacerlo. No voy a repetirlo.\n",
    "\n",
    "Extract the sequence of the closest homolog of G3T0S8_LOXAF in HUMAN and save it a new file called `G3T0S8_best_human_hit.faa`. \n",
    "\n",
    "There are many different ways to do this. You can open the FASTA file containing all proteomes, search by the sequence name and extract it manually, or you can (as you should) do it from the command line. \n",
    "\n",
    "```\n",
    "$ grep [HUMAN_homolog_in_blast_result] all_species.parsed.fasta -A1 > G3T0S8_best_human_hit.faa\n",
    "```\n",
    "\n",
    "#### 5. Now, BLAST G3T0S8_best_human_hit.faa against the same databases. \n",
    "```\n",
    "$ blastp -task blastp -query [HUMAN_seq_file] -db 4proteomes.blastdb -outfmt 6 -evalue 0.001\n",
    "```\n",
    "\n",
    "#### 6. Answer these questions\n",
    "1. Are they reciprocal hits?\n",
    "\n",
    "No. G3T0S8 aparece en tercer lugar con e-value = 1.89e-155\t\n",
    "\n",
    "\n",
    "2. Are they orthologous with each other?\n",
    "\n",
    "O bien TPH1_HUMAN y G3T0S8 no son ortólogos o bien hay una relación de ortología complicada (ej one to many; many to many, etc.).\n",
    "\n",
    "## Goal 2 (Zebrafish orthologs)\n",
    "Repeat the previous protocol with the Zebrafish sequence found in the file `TPH1A_rerio.faa` (Danio rerio homolog)\n",
    "###  Answer these questions: \n",
    "\n",
    "1. What are the Zebrafish homologs in human? \n",
    "2. Are the same as in the Loxodonta example?\n",
    "3. What is the ortholog in human (based on reciprocal blast)?\n",
    "4. What could you tell about the gene TPH1B in Danio rerio?\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
