{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Environment:\n",
    "\n",
    "Login into the GDAV server\n",
    "\n",
    "`ssh youruser@IP`\n",
    "\n",
    "Create a directory in your home folder called `compgenomics_ex2`\n",
    "\n",
    "`mkdir compgenomics_ex2`\n",
    "\n",
    "and Enter the directory \n",
    "\n",
    "`cd compgenomics_ex2`\n",
    "\n",
    "All the files needed for this exercise are copied in the GDAV server at `/home/compgenomics/4proteomes/`. Make sure you can see them and take a few seconds to understand what they contain: \n",
    "\n",
    "```\n",
    "ls  /home/compgenomics/4proteomes/\n",
    "$ ls /home/compgenomics/4proteomes/\n",
    "4proteomes.faa    -> all protein sequences from 4 species: human, elefant, Zebrafish and Ciona intestinalis. \n",
    "G3T0S8_LOXAF.faa  -> protein sequence of the elefant gene called G3T0G8\n",
    "TPH1A_rerio.faa   ->  protein sequence of the Zebrafish gene called TPH1A\n",
    "TPH2_human.faa    -> protein sequence of Human TPH2 \n",
    "scripts/          ->  a directory with ad hoc programs and scripts`\n",
    "```\n",
    "\n",
    "## Tools needed\n",
    "(already installed in the GDAV server)\n",
    "\n",
    "- BLAST+\n",
    "- IQ-Tree\n",
    "- MAFFT\n",
    "- BioPython\n",
    "- ete3\n",
    "- scripts/extract_seqs_from_blast_result.py\n",
    "- scripts/midpoint_rooting.py\n",
    "\n",
    "# Exercise\n",
    "\n",
    "## Goal 1\n",
    "\n",
    "Reconstruct the phylogeny of all TPH2 proteins in the 4 target proteomes, and interpret the tree to identify orthologs. \n",
    "\n",
    "### 1. Identify TPH homologs\n",
    "As in exercise 1, use BLAST to identify all significant hits of the query protein `TPH2_human.faa` in the 4proteomes dataset. \n",
    "\n",
    "Tip: You can reuse the BLAST database of exercise 1 and use a command similar to:\n",
    "\n",
    "```\n",
    "$ blastp -task blastp -query /home/compgenomics/ex1/TPH2_human.faa -db ~/compgenomics_ex1/4proteomes.blastdb -outfmt 6 -evalue 0.001 > TPH2_homologs.blastout\n",
    "```\n",
    "\n",
    "### 2. Extract all homologs in FASTA format\n",
    "\n",
    "Extract all homologs of the TPH2 sequences using the script provided in `/home/compgenomics/ex1/scripts/extract_seqs_from_blast_result.py`. \n",
    "\n",
    "```\n",
    "$ python /home/compgenomics/ex1/scripts/extract_seqs_from_blast_result.py TPH2_homologs.blastout /home/compgenomics/ex1/4proteomes.faa > TPH2_homologs.faa\n",
    "```\n",
    "\n",
    "### 3. Multiple Sequence Alignment (MSA)\n",
    "Before inferring a phylogenetic tree, homologous sequences need to be aligned. There are multiple programs to do it: ClustalOmega, MAFFT, MUSLE, etc. Here, we will use MAFFT, which has a very simple command line. \n",
    "\n",
    "```\n",
    "$ mafft TPH2_homologs.faa > TPH2_homologs.alg\n",
    "```\n",
    "\n",
    "Check the content of the output (saved in TPH2_homologs.alg). What's the main difference compared to the input FASTA file?\n",
    "\n",
    "### 4. Phylogenetic Reconstruction\n",
    "Similarly to MSA programs, there are many software to build phylogenetic trees: RAXML, IQ-TREE, PhyML, FastTree, MrBayes, PhyloBayes, etc. Here we will use IQ-Tree, which uses a Maximum Likelihood approximation. \n",
    "\n",
    "You only need to provide the MSA file as input, and some parameters defining how exhaustive should be the inference. To get a fast result, the following arguments are recommended (avoiding the step of model testing, which is very slow).\n",
    "\n",
    "```\n",
    "$ iqtree -s TPH2_homologs.alg -m LG\n",
    "```\n",
    "\n",
    "### 5. Visualize tree\n",
    "\n",
    "Main IQ-Tree output is the file ending with the `.treefile` extension. The tree file is in [https://en.wikipedia.org/wiki/Newick_format](Newick Format). \n",
    "\n",
    "You can use the command line tool `ete3` to display the directly in the terminal: \n",
    "\n",
    "```\n",
    "$ ete3 view --text -t TPH2_homologs.alg.treefile \n",
    "\n",
    "   /-A0A0R4ILE6_DANRE_tph2\n",
    "  |\n",
    "  |                  /-A0A2R8RPJ0_DANRE_th\n",
    "  |               /-|\n",
    "  |              |  |   /-TY3H_HUMAN_TH\n",
    "  |            /-|   \\-|\n",
    "  |           |  |      \\-G3U1E7_LOXAF_TH\n",
    "  |         /-|  |\n",
    "  |        |  |   \\-Q1LWZ5_DANRE_th2\n",
    "  ...\n",
    "```\n",
    "\n",
    "### 5. Root the tree\n",
    "\n",
    "By default, phylogenetic trees returned by almost all programs are UNROOTED. There are many methods to root a tree, but a common one is the midpoint_rooting. For convenience, an *ad hoc* script to root Newick trees is provided in `/home/compgenomics/ex1/scripts/midpoint_rooting.py` \n",
    "\n",
    "You can use to root your tree before visualizing it: \n",
    "\n",
    "```\n",
    "$ python /home/compgenomics/ex1/scripts/midpoint_rooting.py TPH2_homologs.alg.treefile | ete3 view --text \n",
    "\n",
    "            /-A0A2R8RPJ0_DANRE_th\n",
    "         /-|\n",
    "        |  |   /-TY3H_HUMAN_TH\n",
    "      /-|   \\-|\n",
    "     |  |      \\-G3U1E7_LOXAF_TH\n",
    "   /-|  |\n",
    "  |  |   \\-Q1LWZ5_DANRE_th2\n",
    "  |  |\n",
    "  |   \\-F6Y7Q5_CIOIN_th\n",
    "  |\n",
    "  |         /-Q7SYH6_DANRE_pah\n",
    "--|      /-|\n",
    "  |     |  |   /-PH4H_HUMAN_PAH\n",
    "  |   /-|   \\-|\n",
    "...\n",
    "```\n",
    "\n",
    "\n",
    "Alternatively, you can print the content of the file directly in the terminal and pasted it into any of the online tree visualization server (or transfer the file and upload it): \n",
    "\n",
    "- (http://etetoolkit.org/treeview)\n",
    "- (http://itol.embl.de)\n",
    "\n",
    "### Questions: \n",
    "\n",
    "- What's the evolutionary relationship between F1R1D3_DANRE_tph1a and Q6IWP4_DANRE_tph1b ?\n",
    "- What's the evolutionary relationship between TPH2_HUMAN_TPH2 and TPH1_HUMAN_TPH1 ?\n",
    "- What's the Zebrafish (Danio Rerio) ortholog(s) of the human sequence `TPH1_HUMAN_TPH1`? \n",
    "- How many duplication events can you identify?\n",
    "- How many putative orthologous groups can you identify?  \n",
    "\n",
    "## Goal 2\n",
    "\n",
    "Repeat the same protocol to find all homologs of the P53 sequence in the 4 target proteomes, build a phylogeny, and identigy orthologs. \n",
    "```\n",
    ">P53_HUMAN_TP53\n",
    "MEEPQSDPSVEPPLSQETFSDLWKLLPENNVLSPLPSQAMDDLMLSPDDIEQWFTEDPGPDEAPRMPEAAPPVAPAPAAPTPAAPAPAPSWPLSSSVPSQKTYQGSYGFRLGFLHSGTAKSVTCTYSPALNKMFCQLAKTCPVQLWVDSTPPPGTRVRAMAIYKQSQHMTEVVRRCPHHERCSDSDGLAP\n",
    "PQHLIRVEGNLRVEYLDDRNTFRHSVVVPYEPPEVGSDCTTIHYNYMCNSSCMGGMNRRPILTIITLEDSSGNLLGRNSFEVRVCACPGRDRRTEEENLRKKGEPHHELPPGSTKRALPNNTSSSPQPKKKPLDGEYFTLQIRGRERFEMFRELNEALELKDAQAGKEPGGSRAHSSHLKSKKGQSTSRH\n",
    "KKLMFKTEGPDSD\n",
    "```\n",
    "\n",
    "### Questions:\n",
    "1. Could you identify duplication and speciation events?\n",
    "2. How many putative orthologous groups can you identify?  \n",
    "3. Is there anything unusual in the evolution of this gene family?\n",
    "4. Is the tree rooted? Where would you root it? \n",
    "5. Upload the tree into http://itol.embl.de and explore rooting options. Does it change the inference of duplication events?\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
